src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs

```diff
@@ -170,7 +170,7 @@ public static partial class JsonSerializer
                 {
                     state.Current.TempEnumerableValues.Add(value);
                 }
-                else
+                else if (state.Current.ReturnValue != null)
```

Comment:
It seems like this is the wrong code path since the value should never be added to a list. It should be getting to the SetValueAsObject code a little lower which is used to set the return value to the root method being called. I'll take a closer look later and verify.

Arrays are different than IList\IEnumerable since they need to create a temporary list to hold that values until all values are read, but in this case it should not have a temporary list nor attempt to add a null to one.

Reply:
I see what you mean. It probably should never reach `ApplyObjectToEnumerable`. 
I'll try to see if there's a clean way to catch root-level nulls within `HandleNull`.

Side note: Looks like this issue encompasses everything of `ClassType.Enumerable`.
The following test throws null-ref as well:
```
IEnumerable<object> obj = JsonSerializer.Parse<IEnumerable<object>>("null");
```

Reply:
> Side note: Looks like this issue encompasses everything of `ClassType.Enumerable`.

@gnovack, while you are investigating this space, can you please add tests for similar cases like the one you mentioned? Similarly, `Parse<T>` where T is different types: **value types** (`int`, `DateTime`, etc.) and **reference types** (`string`, some POCO class, `Dictionary<TKey, TValue>`, and as you mentioned `IEnumerable<T>`).